File: packages/typescript-semantic-match/src/semanticItem.ts
/**
 * Shared item primitives and helpers used by semantic list comparison.
 *
 * This module intentionally focuses on item-level behavior:
 * - `SemanticItem` defines the accepted item shape (`string` or `{ name, description? }`).
 * - `getItemName` normalizes an item to its comparable name.
 * - `itemToPromptString` formats an item for prompt text, including optional details.
 * - `compareItems` provides case-insensitive ordering by item name.
 * - `areItemsEqual` provides equality checks across comparable item content.
 *
 * Matching orchestration (removed/added/renamed classification) is implemented in
 * higher-level modules and consumes these utilities.
 */

/**
 * Item shape accepted by `compareItemLists` for semantic comparison.
 *
 * - A raw string is treated as the item's comparable name.
 * - An object uses `name` as the comparable value and may include optional
 *   `description` to provide additional LLM context.
 */
export type SemanticItem =
  | string
  | { name: string; description?: string };


/**
 * Returns the comparable name for a list item.
 * @param item The item to extract the name from.
 * @returns The name of the item, which is used for comparison and matching.
 */
export const getItemName = (item: SemanticItem): string => {
  return typeof item === 'string' ? item : item.name;
};

/**
 * Returns the description of a list item, if available and non-redundant with the name.
 * If the item is a string or if the description is missing or effectively the same as the name,
 * this function returns `undefined`.
 * @param item The item to extract the description from.
 * @returns The description of the item, or `undefined` if not available or redundant.
 */
export const getItemDescription = (item: SemanticItem): string | undefined => {
  if (typeof item === 'string') {
    return undefined;
  }
  if (!item.description) {
    return undefined;
  }
  // If the description is the same as the name (ignoring case and whitespace),
  // then it's not really providing any additional context, so we can ignore it.
  if (item.description.trim().toLowerCase() === item.name.trim().toLowerCase()) {
    return undefined;
  }
  return item.description;
}


/**
 * Formats a list item for prompt inclusion, including optional description context.
 * The output is a string that starts with "- " followed by the item name, and if a 
 * description is provided and is not redundant with the name, it includes the description 
 * in parentheses. The item name and description are JSON-stringified to prevent formatting
 * issues in the prompt (e.g. with newlines or special characters).
 * @param item The item to format for the prompt.
 * @returns A string representation of the item suitable for inclusion in the prompt.
 */
export const itemToPromptString = (item: SemanticItem): string => {
  let s = `- ${JSON.stringify(getItemName(item))}`;
  const description = getItemDescription(item);
  if (description) {
    s += ` (details: ${JSON.stringify(description)})`;
  }
  return s;
};

/**
 * Sort comparator for list items.
 *
 * Ordering behavior:
 * 1) Compare names case-insensitively after trimming leading/trailing whitespace.
 * 2) If names are equal, compare non-redundant descriptions case-insensitively
 *    as a tie-breaker. We only compare descriptions when both items have a
 *    non-redundant description.
 */
export const compareItems = (
  a: SemanticItem,
  b: SemanticItem
) => {
  const nameA = getItemName(a).trim().toLowerCase();
  const nameB = getItemName(b).trim().toLowerCase();
  const byName = nameA.localeCompare(nameB);
  if (byName !== 0) {
    return byName;
  }

  const descA = (getItemDescription(a) ?? '').trim().toLowerCase();
  const descB = (getItemDescription(b) ?? '').trim().toLowerCase();

  // In order to compare descriptions, both items should have a description.
  // If only one item has a description, we don't bother comparing the
  // description field.
  if (!descA || !descB) {
    return 0;
  }

  // If we have two descriptions, we can use them as a tie-breaker.
  return descA.localeCompare(descB);
};

/**
 * Equality check for two items.
 *
 * Equality uses the same semantics as `compareItems`:
 * - names are compared case-insensitively after trimming;
 * - when names tie, non-redundant descriptions are compared
 *   case-insensitively after trimming.
 * @param a The first item to compare.
 * @param b The second item to compare.
 * @returns `true` if the items are equal under comparator semantics, `false` otherwise.
 */
export const areItemsEqual = (a: SemanticItem, b: SemanticItem): boolean => {
  return compareItems(a, b) === 0;
}

/**
 * Removes an item from a list based on full item equivalence.
 * @param itemList The list of items to remove from.
 * @param itemToRemove The item to remove from the list. Any item equal to this item
 * under `areItemsEqual` semantics will be removed.
 * @returns A new list with the specified item removed.
 */
export const removeItemFromList = (itemList: SemanticItem[], itemToRemove: SemanticItem): SemanticItem[] => {
  return itemList.filter(item => !areItemsEqual(item, itemToRemove));
}

File: packages/python-semantic-match/src/mightydatainc_semantic_match/semantic_item.py
"""Shared item primitives and helpers for semantic list comparison.

This module focuses on item-level behavior:

- ``SemanticItem`` defines the accepted item shape (``str`` or
  ``{"name", "description"?}``).
- ``get_item_name`` normalizes an item to its comparable name.
- ``get_item_description`` returns a non-redundant optional description.
- ``item_to_prompt_string`` formats an item for prompt text.
- ``compare_items`` provides case-insensitive ordering by item content.
- ``are_items_equal`` checks comparator-based equality.
- ``remove_item_from_list`` removes all equivalent items from a list.
"""

from __future__ import annotations

import json
from typing import NotRequired, TypeAlias, TypedDict


class ComparableNamedItem(TypedDict):
    """Object-form list item used for semantic comparison."""

    name: str
    description: NotRequired[str]


SemanticItem: TypeAlias = str | ComparableNamedItem
"""Accepted item input type for semantic comparison utilities."""


def get_item_name(item: SemanticItem) -> str:
    """Return the comparable name for an item."""
    if isinstance(item, str):
        return item
    return item["name"]


def get_item_description(item: SemanticItem) -> str | None:
    """Return a non-redundant description for an item.

    Returns ``None`` when the item is a string, has no description,
    or when the description is equivalent to the name after trimming and
    case normalization.
    """
    if isinstance(item, str):
        return None

    description = item.get("description")
    if not description:
        return None

    if description.strip().lower() == item["name"].strip().lower():
        return None

    return description


def item_to_prompt_string(item: SemanticItem) -> str:
    """Format an item for prompt inclusion.

    The output begins with ``- `` and JSON-quoted name text, with optional
    ``(details: ...)`` when a non-redundant description is present.
    """
    prompt_item = f"- {json.dumps(get_item_name(item))}"
    description = get_item_description(item)
    if description:
        prompt_item += f" (details: {json.dumps(description)})"
    return prompt_item


def compare_items(a: SemanticItem, b: SemanticItem) -> int:
    """Comparator for semantic items.

    Comparison order:
    1) Names, case-insensitive after trimming.
    2) If names tie and both items have non-redundant descriptions,
       compare descriptions case-insensitively after trimming.
    """
    name_a = get_item_name(a).strip().lower()
    name_b = get_item_name(b).strip().lower()
    if name_a < name_b:
        return -1
    if name_a > name_b:
        return 1

    description_a = (get_item_description(a) or "").strip().lower()
    description_b = (get_item_description(b) or "").strip().lower()

    if not description_a or not description_b:
        return 0

    if description_a < description_b:
        return -1
    if description_a > description_b:
        return 1
    return 0


def are_items_equal(a: SemanticItem, b: SemanticItem) -> bool:
    """Return whether two items are equal under ``compare_items`` semantics."""
    return compare_items(a, b) == 0


def remove_item_from_list(
    item_list: list[SemanticItem],
    item_to_remove: SemanticItem,
) -> list[SemanticItem]:
    """Return a new list with all equivalent items removed."""
    return [item for item in item_list if not are_items_equal(item, item_to_remove)]
